### DISCUSSION TITLE
Expose the number of retries performed in a response

### DISCUSSION THREAD

**Discussion by koniiiik** [encode/httpx#2928] [Ideas]
Something like an integer `Response.retries` that would reveal how many times the transport had to retry the connection before succeeding. It would have to be implemented by `httpcore.HTTPConnection._connect`, and `httpcore.AsyncHTTPConnection._connect`, with matching plumbing up the stack.

The reason why I'm looking for this is that we're trying to use the reconnect feature as a workaround for random connection errors that we can't quite explain yet, and want to gather some data, see if there are any patterns to them.

It might potentially be interesting to also get the exception instances that got raised during the retry loop, but it might also be too much unnecessary detail…? On the other hand, if retries for specified HTTP status codes get implemented, this level of detail might also become more interesting…?

Ideas, opinions?

**Comment by lovelydinosaur** [comment]
This?... `len(response.history)`

**Reply by koniiiik** [reply]
I don't think so… I may be wrong, but doesn't `response.history` only contain the redirect chain? As far as I can tell, connection retries don't generate any response object, in fact, they are currently completely invisible to anything higher up the stack than `httpcore.HTTPConnection._connect`

**Reply by lovelydinosaur** [reply]
Ah sorry not reading your question carefully enough there.

The closest you'd currently be able to get to this would be using [the `trace` request extension](https://www.encode.io/httpcore/extensions/#trace) to monitor connection attempts.

I don't imagine that we'd want this as a built-in functionality, but there is more that we could do to allow custom transport backends so that you're able to customise the instrumentation at that kind of level. If you're interested in working through that then I'd suggest taking a look at working directly with an `httpcore.Connection()` instance as a start. You'd be able to work from there, to then using a custom `httpcore.ConnectionPool()` class, and from there to using a custom `httpx.Transport` class.
